File: Commons/Action/src/main/java/com/greentree/action/observable/RunObservable.java
package com.greentree.action.observable;

import java.io.Serializable;

import com.greentree.action.ListenerCloser;

public interface RunObservable extends Serializable {

	RunObservable NULL = new RunObservable() {
		private static final long serialVersionUID = 1L;

		@Override
		public int listenerSize() {
			return 0;
		}
		
		@Override
		public ListenerCloser addListener(Runnable listener) {
			return ListenerCloser.EMPTY;
		}
	};
	
	ListenerCloser addListener(Runnable listener);
	int listenerSize();
	
	default boolean isEmpty() {
		return listenerSize() == 0;
	}
	
}

File: Commons/Event/src/main/java/com/greentree/commons/event/observable/RunObservable.kt
package com.greentree.commons.event.observable

import com.greentree.commons.event.ListenerCloser
import java.io.Serializable

interface RunObservable : AutoCloseable, Serializable {

	fun addListener(listener: () -> Unit): ListenerCloser

	data object Empty : RunObservable {

		private fun readResolve(): Any = Empty

		override fun addListener(listener: () -> Unit) = ListenerCloser.Empty

		override fun close() {
		}
	}

	companion object {

		@Deprecated("use object", ReplaceWith("RunObservable.Empty"))
		@JvmField
		val NULL = Empty
	}
}

@Deprecated("", ReplaceWith("addListener { listener.run() }"))
fun RunObservable.addListener(listener: Runnable): ListenerCloser = addListener { listener.run() }
